File: platforms/ios/Dezel/Dezel/View/Style/Stylesheet.swift
/**
 * @class Stylesheet
 * @since 0.7.0
 */
open class Stylesheet {

	//--------------------------------------------------------------------------
	// MARK: Properties
	//--------------------------------------------------------------------------

	/**
	 * @property delegate
	 * @since 0.7.0
	 */
	open weak var delegate: StylesheetDelegate?

	/**
	 * @property handle
	 * @since 0.7.0
	 * @hidden
	 */
	internal var handle: StylesheetRef

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.7.0
	 */
	public init() {
		self.handle = StylesheetCreate()
	}

	/**
	 * @destructor
	 * @since 0.7.0
	 */
	deinit {
		StylesheetDelete(self.handle)
	}

	/**
	 * @method setVariable
	 * @since 0.7.0
	 */
	public func setVariable(_ name: String, value: String) {

		var error: UnsafeMutablePointer<ParseError>?

		StylesheetSetVariable(self.handle, name, value, &error)

		if let error = error {
			self.delegate?.didThrowError(
				stylesheet: self,
				error: error.pointee.message.string,
				col: Int(error.pointee.col),
				row: Int(error.pointee.row),
				url: error.pointee.url.string
			)
		}
	}

	/**
	 * @method evaluate
	 * @since 0.7.0
	 */
	public func evaluate(_ source: String, url: String) {

		var error: UnsafeMutablePointer<ParseError>?

		StylesheetEvaluate(self.handle, source, url, &error)

		if let error = error {
			self.delegate?.didThrowError(
				stylesheet: self,
				error: error.pointee.message.string,
				col: Int(error.pointee.col),
				row: Int(error.pointee.row),
				url: error.pointee.url.string
			)
		}
	}
}

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/view/display/Stylesheet.kt
package ca.logaritm.dezel.view.display

import android.util.Log
import ca.logaritm.dezel.view.display.exception.ParseError
import ca.logaritm.dezel.view.display.external.StylesheetExternal

/**
 * @class Stylesheet
 * @since 0.7.0
 */
open class Stylesheet {

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * @property listener
	 * @since 0.7.0
	 */
	open var listener: StylesheetListener? = null

	/**
	 * @property handle
	 * @since 0.7.0
	 */
	public var handle: Long = 0L
		private set

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.7.0
	 */
	init {
		this.handle = StylesheetExternal.create()
	}

	/**
	 * @method setVariable
	 * @since 0.7.0
	 */
	open fun setVariable(name: String, value: String) {

		try {

			StylesheetExternal.setVariable(this.handle, name, value)

		} catch (e: ParseError) {

			this.listener?.onThrowError(
				stylesheet = this,
				error = e.message,
				col = e.col,
				row = e.row,
				url = e.url
			)

		}
	}

	/**
	 * @method evaluate
	 * @since 0.7.0
	 */
	open fun evaluate(source: String, url: String) {

		try {

			StylesheetExternal.evaluate(this.handle, source, url)

		} catch (e: ParseError) {

			this.listener?.onThrowError(
				stylesheet = this,
				error = e.message,
				col = e.col,
				row = e.row,
				url = e.url
			)

		}
	}
}